src/evaluateXPath.ts

```diff
@@ -136,6 +147,112 @@ const evaluateXPath = <TNode extends Node, TReturnType extends keyof IReturnType
 		throw new TypeError("Failed to execute 'evaluateXPath': xpathExpression must be a string.");
 	}
 
+	const backend = options && options.backend ? options.backend : 'expression';
+
+	if (backend === 'expression') {
+		return evaluateWithExpressionBackend(
+			selector,
+			contextItem,
+			domFacade,
+			variables,
+			returnType,
+			options
+		);
+	} else if (backend === 'js-codegen') {
+		return evaluateWithJsCodegenBackend(selector, contextItem, domFacade, returnType, options);
+	} else if (backend === 'auto') {
+		try {
+			return evaluateWithJsCodegenBackend(
```

Comment:
> Hm, this try the code gen, if it breaks -> just use the expression backend feels silly.

I agree. In some cases errors are thrown because an AST can not be compiled because it is unsupported, while in other times it may throw for different reasons.

> Instead, how would you feel to expose a function that scans an AST for feasibility for codegen, but returns something that it does not work. Otherwise, if an expression breaks during execution, it will just try again in the Expression backend. Which feels silly.

I don't think this is the best option. If I understand correctly, this would mean two functions have to be maintained: one that scans the AST and one that transforms it into JS code. Since I think the js codegen backend should throw errors when it can't compile an AST, I feel like this is duplication of the same idea.

What if we make a distinction between errors? If the thrown error is of type `UnsupportedAST`, then try compiling the AST with the expression backend. If the error is of a different type, then it should be handled differently (maybe just rethrow the error, which is what happens when the expression backend encounters an error).

Reply:
Not when you make the API something like this:

```
function compileToJavaScript(ast: IAst): {astIsAcceptable: true, result: string?}|{astIsAcceptable: false}
```

That boolean can just be returned instead of an error.

Later check whether the AST is acceptable, in the codegen branch just throw if it isn't but in the automatic flow, just fall back to JavaScript.

This is also a bit more usable in the future when we pull this parse + compile to the build phase of the Fonto editor: we can just scan the JavaScript (or Typescript at that point) and replace any and all XPaths with the compiled JavaScript counterpart.

Reply:
Do you think it would be a good idea to specify a message with `{astIsAcceptable: false}`? Like `{astIsAcceptable: false, reason: string}`? This can make it easier to see why the AST is not accepted, which can make debugging easier. Maybe if the reason is supplied one can decide to implement the feature in the backend.

Reply:
Done